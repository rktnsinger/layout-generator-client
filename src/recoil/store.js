import { atom } from "recoil";
import { holy, holySize } from "../mockLayout";

const imageURLState = atom({
  key: "imageURLState",
  default: "",
});

const layoutTypeState = atom({
  key: "layoutTypeState",
  default: "",
});

const imageSizeState = atom({
  key: "imageSizeState",
  default: holySize,
});

const detectedLinesState = atom({
  key: "detectedLines",
  default: holy,
});

const generatedCodeState = atom({
  key: "generatedCode",
  default: "",
});

export {
  imageURLState,
  imageSizeState,
  layoutTypeState,
  detectedLinesState,
  generatedCodeState,
};
