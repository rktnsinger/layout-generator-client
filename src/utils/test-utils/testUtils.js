/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import { BrowserRouter } from "react-router-dom";
import PropTypes from "prop-types";
import { ThemeProvider } from "styled-components";
import { render } from "@testing-library/react";

import theme from "../../styles/theme";

function CustomWrapper({ children }) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (ui, options) =>
  render(ui, { wrapper: CustomWrapper, ...options });

export * from "@testing-library/react";
export { customRender as render };

CustomWrapper.propTypes = {
  children: PropTypes.node.isRequired,
};
