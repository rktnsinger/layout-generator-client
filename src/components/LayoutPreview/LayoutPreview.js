/* eslint-disable react/no-array-index-key */
import React from "react";
import { useRecoilValue } from "recoil";
import { Navigate } from "react-router-dom";

import { detectedLinesState, imageSizeState } from "../../recoil/store";
import {
  getFractionString,
  getLayoutType,
  isProperLineData,
} from "../../utils/layoutUtils";
import isInRange from "../../utils/isInRange";

import { DIVISION_TYPE, LAYOUT_TYPE, ERROR } from "../../constants";

export default function LayoutPreview() {
  const imageSize = useRecoilValue(imageSizeState);
  const detectedLines = useRecoilValue(detectedLinesState);

  const { width: imageWidth, height: imageHeight } = imageSize;
  const { rowLines, columnLines } = detectedLines;
  const isProperLine = isProperLineData(rowLines, columnLines);

  if (!isProperLine) {
    return <Navigate to="/error" state={ERROR.invalidLine} />;
  }

  const layoutType = getLayoutType(rowLines, columnLines, imageWidth);
  const gridTemplateRows = getFractionString(rowLines, imageHeight);
  const gridTemplateColumns = getFractionString(columnLines, imageWidth);

  const divisions = [];
  let currentDivision = DIVISION_TYPE.full;
  let currentLineIndex = 0;
  let intersection = 0;
  let intersectionIndex = 0;

  switch (layoutType) {
    case LAYOUT_TYPE.row:
      return (
        <section
          style={{
            display: "grid",
            gridTemplateRows,
          }}
        >
          {Array.from({ length: rowLines.length + 1 }).map((item, index) => (
            <div key={index}>{index + 1}</div>
          ))}
        </section>
      );

    case LAYOUT_TYPE.column:
      return (
        <section
          style={{
            display: "grid",
            gridTemplateColumns,
          }}
        >
          {Array.from({ length: columnLines.length + 1 }).map((item, index) => (
            <div key={index}>{index + 1}</div>
          ))}
        </section>
      );

    case LAYOUT_TYPE.rowC:
      while (divisions.length < rowLines.length + columnLines.length + 1) {
        intersection = columnLines[0].endY;

        if (currentDivision === DIVISION_TYPE.full) {
          divisions.push(DIVISION_TYPE.full);

          if (isInRange(rowLines[currentLineIndex].startY, intersection, 20)) {
            currentDivision = DIVISION_TYPE.partial;
            intersectionIndex = currentLineIndex;
            currentLineIndex = 0;
          } else {
            currentLineIndex += 1;
          }
        }

        if (currentDivision === DIVISION_TYPE.partial) {
          if (columnLines.length - 1 === currentLineIndex) {
            divisions.push(DIVISION_TYPE.partial, DIVISION_TYPE.partial);

            currentDivision = DIVISION_TYPE.full;
            currentLineIndex = intersectionIndex + 1;
          } else {
            divisions.push(DIVISION_TYPE.partial);

            currentLineIndex += 1;
          }
        }
      }

      return (
        <section
          style={{
            display: "grid",
            gridTemplateRows,
            gridTemplateColumns,
          }}
        >
          {divisions.map((div, index) => {
            return div === DIVISION_TYPE.full ? (
              <div
                key={`${div}-${index}`}
                style={{ gridColumn: `1 / ${columnLines.length + 2}` }}
              >
                {index + 1}
              </div>
            ) : (
              <div key={`${div}-${index}`}>{index + 1}</div>
            );
          })}
        </section>
      );

    default:
      return <Navigate to="/error" state={ERROR.notSupport} />;
  }
}
