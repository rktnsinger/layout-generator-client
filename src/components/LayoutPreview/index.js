import LayoutPreview from "./LayoutPreview";

export default LayoutPreview;
