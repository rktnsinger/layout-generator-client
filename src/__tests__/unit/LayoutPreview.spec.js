import React from "react";
import { RecoilRoot } from "recoil";
import { Navigate } from "react-router-dom";

import LayoutPreview from "../../components/LayoutPreview";

import { detectedLinesState, imageSizeState } from "../../recoil/store";
import { render } from "../../utils/test-utils/testUtils";

jest.mock("react-router-dom", () => {
  const originalModule = jest.requireActual("react-router-dom");

  return {
    ...originalModule,
    Navigate: jest.fn(),
  };
});

describe("<LayoutPreview />", () => {
  const IMAGE_SIZE = { width: 500, height: 300 };

  const renderSnapshot = (lines) => {
    const initializeState = ({ set }) => {
      set(imageSizeState, IMAGE_SIZE);
      set(detectedLinesState, lines);
    };

    const { baseElement } = render(
      <RecoilRoot initializeState={initializeState}>
        <LayoutPreview />
      </RecoilRoot>
    );

    return baseElement;
  };

  describe("01-LAYOUT_TYPE.row(rowLine만 존재하는 경우)", () => {
    it("case: 1 rowLine", () => {
      const detectedLines = {
        rowLines: [{ startX: 0, startY: 100, endX: 500, endY: 100 }],
        columnLines: [],
      };
      const view = renderSnapshot(detectedLines);

      expect(view).toMatchSnapshot();
    });

    it("case: 2 rowLine", () => {
      const detectedLines = {
        rowLines: [
          { startX: 0, startY: 50, endX: 500, endY: 50 },
          { startX: 0, startY: 250, endX: 500, endY: 250 },
        ],
        columnLines: [],
      };
      const view = renderSnapshot(detectedLines);

      expect(view).toMatchSnapshot();
    });
  });

  describe("02-LAYOUT_TYPE.column(columnLine만 존재하는 경우)", () => {
    it("case: 1 columnLine", () => {
      const detectedLines = {
        rowLines: [],
        columnLines: [{ startX: 150, startY: 300, endX: 150, endY: 0 }],
      };
      const view = renderSnapshot(detectedLines);

      expect(view).toMatchSnapshot();
    });

    it("case: 2 columnLines", () => {
      const detectedLines = {
        rowLines: [],
        columnLines: [
          { startX: 100, startY: 300, endX: 100, endY: 0 },
          { startX: 400, startY: 300, endX: 400, endY: 0 },
        ],
      };
      const view = renderSnapshot(detectedLines);

      expect(view).toMatchSnapshot();
    });
  });

  describe("03-LAYOUT_TYPE.complexedRow", () => {
    it("case: 1 rowLine, 1 columnLine", () => {
      const detectedLines = {
        rowLines: [{ startX: 0, startY: 80, endX: 500, endY: 80 }],
        columnLines: [{ startX: 100, startY: 300, endX: 100, endY: 80 }],
      };
      const view = renderSnapshot(detectedLines);

      expect(view).toMatchSnapshot();
    });

    it("case: 1 rowLine, 2 columnLine", () => {
      const detectedLines = {
        rowLines: [{ startX: 0, startY: 80, endX: 500, endY: 80 }],
        columnLines: [
          { startX: 100, startY: 300, endX: 100, endY: 80 },
          { startX: 380, startY: 300, endX: 380, endY: 80 },
        ],
      };
      const view = renderSnapshot(detectedLines);

      expect(view).toMatchSnapshot();
    });

    it("case: 2 rowLine, 2 columnLine", () => {
      const detectedLines = {
        rowLines: [
          { startX: 0, startY: 80, endX: 500, endY: 80 },
          { startX: 0, startY: 220, endX: 500, endY: 220 },
        ],
        columnLines: [
          { startX: 100, startY: 220, endX: 100, endY: 80 },
          { startX: 400, startY: 220, endX: 400, endY: 80 },
        ],
      };
      const view = renderSnapshot(detectedLines);

      expect(view).toMatchSnapshot();
    });
  });

  describe("04-default(사전 정의된 layoutType이 아닌 경우)", () => {
    it("에러페이지로 리다이렉트 되어야 한다", () => {
      const detectedLines = {
        rowLines: [
          { startX: 80, startY: 100, endX: 350, endY: 100 },
          { startX: 80, startY: 400, endX: 350, endY: 400 },
        ],
        columnLines: [
          { startX: 80, startY: 300, endX: 80, endY: 0 },
          { startX: 350, startY: 300, endX: 350, endY: 0 },
        ],
      };

      const initializeState = ({ set }) => {
        set(imageSizeState, IMAGE_SIZE);
        set(detectedLinesState, detectedLines);
      };

      render(
        <RecoilRoot initializeState={initializeState}>
          <LayoutPreview />
        </RecoilRoot>
      );

      expect(Navigate).toBeCalled();
    });
  });
});
