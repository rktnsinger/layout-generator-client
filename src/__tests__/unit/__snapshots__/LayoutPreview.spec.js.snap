// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LayoutPreview /> 01-LAYOUT_TYPE.row(rowLine만 존재하는 경우) case: 1 rowLine 1`] = `
<body>
  <div>
    <section
      style="display: grid; grid-template-rows: 3.3fr 6.7fr;"
    >
      <div>
        1
      </div>
      <div>
        2
      </div>
    </section>
  </div>
</body>
`;

exports[`<LayoutPreview /> 01-LAYOUT_TYPE.row(rowLine만 존재하는 경우) case: 2 rowLine 1`] = `
<body>
  <div>
    <section
      style="display: grid; grid-template-rows: 1.7fr 8.3fr 1.7fr;"
    >
      <div>
        1
      </div>
      <div>
        2
      </div>
      <div>
        3
      </div>
    </section>
  </div>
</body>
`;

exports[`<LayoutPreview /> 02-LAYOUT_TYPE.column(columnLine만 존재하는 경우) case: 1 columnLine 1`] = `
<body>
  <div>
    <section
      style="display: grid; grid-template-columns: 3.0fr 7.0fr;"
    >
      <div>
        1
      </div>
      <div>
        2
      </div>
    </section>
  </div>
</body>
`;

exports[`<LayoutPreview /> 02-LAYOUT_TYPE.column(columnLine만 존재하는 경우) case: 2 columnLines 1`] = `
<body>
  <div>
    <section
      style="display: grid; grid-template-columns: 2.0fr 8.0fr 2.0fr;"
    >
      <div>
        1
      </div>
      <div>
        2
      </div>
      <div>
        3
      </div>
    </section>
  </div>
</body>
`;

exports[`<LayoutPreview /> 03-LAYOUT_TYPE.complexedRow case: 1 rowLine, 1 columnLine 1`] = `
<body>
  <div>
    <section
      style="display: grid; grid-template-rows: 2.7fr 7.3fr; grid-template-columns: 2.0fr 8.0fr;"
    >
      <div
        style="grid-column: 1 / 3;"
      >
        1
      </div>
      <div>
        2
      </div>
      <div>
        3
      </div>
    </section>
  </div>
</body>
`;

exports[`<LayoutPreview /> 03-LAYOUT_TYPE.complexedRow case: 1 rowLine, 2 columnLine 1`] = `
<body>
  <div>
    <section
      style="display: grid; grid-template-rows: 2.7fr 7.3fr; grid-template-columns: 2.0fr 7.6fr 2.4fr;"
    >
      <div
        style="grid-column: 1 / 4;"
      >
        1
      </div>
      <div>
        2
      </div>
      <div>
        3
      </div>
      <div>
        4
      </div>
    </section>
  </div>
</body>
`;

exports[`<LayoutPreview /> 03-LAYOUT_TYPE.complexedRow case: 2 rowLine, 2 columnLine 1`] = `
<body>
  <div>
    <section
      style="display: grid; grid-template-rows: 2.7fr 7.3fr 2.7fr; grid-template-columns: 2.0fr 8.0fr 2.0fr;"
    >
      <div
        style="grid-column: 1 / 4;"
      >
        1
      </div>
      <div>
        2
      </div>
      <div>
        3
      </div>
      <div>
        4
      </div>
      <div
        style="grid-column: 1 / 4;"
      >
        5
      </div>
    </section>
  </div>
</body>
`;
